### PR TITLE
Feature 4: Users can filter spells list

### LINKED ISSUE #4: Feature 4: User can search on the spell list and filter by various criteria.
## ⚙️ Feature

> ##### What capability will users have when this feature is merged?

User can search for partial or whole terms and the spells list will filter to include spells that match the search term. The user can also fill out an advanced filter to search by additional criteria.

## 📐 Functionality

> ##### A description of how the application will work from the perspective of the user.

The user can search spells by filling out a search bar at the top of the page. If they click on the advanced filter button, it will open a modal with a form for the additional criteria, then filter the spells list and update it.

## 🖼️ Wireframes

> ##### Wireframe screenshot(s) of what the app should look like when it is complete.

image

## ✅ Task List

> ##### Commit checklist for tasks needed to complete the feature.

- [x] Create a new object to store sort/search/filter information.
- [x] Create a form object for the search bar.
- [x] Create an input element and a submit button within the form.
- [x] when the form is submitted, store the values in the sort object
- [x] Create a button that will open the form modal.
- [x] Create a dialog with a form.
  - [x] The form has 3 options: name, level, school
  - [x] name is an input
  - [x] level is a dropdown
  - [x] school is a dropdown
- [x] create a new array to store the list of filtered cards
- [x] create a new div to store the filtered list of cards
- [x] create a function to filter the card array
  - [x] query the api using the appropriate filters
  - [x] render cards and append them to the filtered card array
  - [x] clear the div of filtered cards
  - [x] append the filtered array to the div
- [x] when the filter form is submitted
  - [x] store the form values
  - [x] if the values are not the default, filter the cards
  - [x] hide the all cards list and show the filtered cards list
- [x] when the filter values are reset
  - [x] hide the filtered list and show the all cards list
- **Clean Up**
  - [x] Remove all buttons and placeholders that do not serve the functionality of the feature.
  - [x] Remove all console logs.
  - [x] Remove all commented out code.
  - [x] Remove all CSS properties that have no effect.
  - [x] Check all code for proper formatting and indentation.
  - [x] Confirm that there are no errors in the console while using the application.
  - [x] Confirm that all previous functionality still works without errors.
  - [x] Confirm that the user interface looks natural on both mobile and desktop screens.
  - [x] Create GIFs of the feature being used on both mobile and desktop screen sizes for inclusion in Pull Request.

### PR DESCRIPTION
**Pull request by TristanChao**
Recording 2024-08-19 at 16 47 33
Recording 2024-08-19 at 16 48 42
Recording 2024-08-19 at 16 49 15
Recording 2024-08-19 at 16 50 32

### CODE REVIEW

File: ts/main.ts [hunk 1]
```diff
@@ -79,6 +135,24 @@ function levelNumberToString(level: number): string {
...
+  } else if (list === 'filtered cards') {
+    if (!$spellsListCardsDiv.className.includes('hidden')) {
+      $spellsListCardsDiv.className += ' hidden';
+    }
+    if ($spellsListFilteredCardsDiv.className.includes('hidden')) {
+      $spellsListFilteredCardsDiv.classList.remove('hidden');
+    }
+  }
+}
+
```
Comment (brandon-moy) [inline ts/main.ts]: Depending on the changes below will it be necessary to have different views for filtered cards and all of the cards? If we can make the changes then we would just need one view for all of the cards. This function would maybe still be needed as your view swap, but for spell details and spell books.

File: ts/main.ts [hunk 2]
```diff
@@ -194,49 +263,187 @@ async function renderAllCardsInitial(): Promise<void> {
...
   // clear cards
   while ($spellsListCardsDiv.childNodes.length > 0) {
     if (!$spellsListCardsDiv.firstElementChild) break;
     $spellsListCardsDiv.removeChild($spellsListCardsDiv.firstElementChild);
   }
 
   // re-render cards based on sort value
-  if ($spellsListSortDropdown.value === 'level') {
-    sortCardsLevel();
+  sortCards(cardSort.sort);
```
Comment (brandon-moy) [inline ts/main.ts]: Was this from the last feature? Would we be able to do this without deleting and recreating all of the cards?

File: ts/main.ts [hunk 3]
```diff
@@ -194,49 +263,187 @@ async function renderAllCardsInitial(): Promise<void> {
...
+      );
+    }
+    // appends cards into the filtered cards div
+    filteredArray.forEach((element) => {
+      $spellsListFilteredCardsDiv.appendChild(element);
+    });
+  } catch (err) {
+    console.error('Error:', err);
+  }
+}
```
Comment (brandon-moy) [inline ts/main.ts]: Is this removing all cards and then fetching data again? Is there any way to tag more attributes to each card that you could check, loop through the list of spell cards, and then conditional show or hide them using CSS instead?

File: ts/main.ts [hunk 4]
```diff
@@ -194,49 +263,187 @@ async function renderAllCardsInitial(): Promise<void> {
...
+      switchCardsList('all cards');
+      return;
+    }
+
+    // if the filter form is not empty, it will filter the spells and
+    // switch to the filtered list
+    await filterSpellsList();
+    switchCardsList('filtered cards');
+  } catch (err) {
+    console.error('Error:', err);
```
Comment (brandon-moy) [inline ts/main.ts]: Same thing here, can we just modify the elements we already have rather than deleting and recreating all of them?

**Review by brandon-moy — CHANGES_REQUESTED** [review]
Few things to look at and change!

**Comment by TristanChao** [comment]
Recording 2024-08-20 at 09 59 02

File: ts/main.ts [hunk 5]
```diff
@@ -194,56 +239,175 @@ async function renderAllCardsInitial(): Promise<void> {
...
+    // loops through the card array and shows or hides depending if its name
+    // is included in the array of names
+    cardsArray.forEach((element: HTMLDivElement) => {
+      const elementName = element.getAttribute('data-name') as string;
+      if (!elementName) return;
+      if (filteredSpellNames.includes(elementName)) {
+        element.classList.remove('hidden');
+      } else {
+        element.className += ' hidden';
+      }
```
Comment (brandon-moy) [inline ts/main.ts]: Minor thing but if you used `classList` on 269 I would use `classList` on 271 for consistency.
`element.classList.add('hidden')`

**Review by brandon-moy — APPROVED** [review]
✅

**Pull request merged**